core/src/bundle_stage.rs

```diff
@@ -122,6 +123,53 @@ struct AllExecutionResults {
     pub post_balances: (TransactionBalances, TransactionTokenBalances),
 }
 
+struct BundleReservedSpace {
+    current_tx_block_limit: u64,
+    current_bundle_block_limit: u64,
+    initial_allocated_cost: u64,
+}
+
+impl BundleReservedSpace {
+    fn reset_reserved_cost(&mut self, working_bank: &Arc<Bank>) {
+        self.current_bundle_block_limit = MAX_BLOCK_UNITS;
+        self.current_tx_block_limit = MAX_BLOCK_UNITS - self.initial_allocated_cost;
+
+        working_bank.write_cost_tracker().unwrap().set_limits(
+            MAX_WRITABLE_ACCOUNT_UNITS,
+            self.current_tx_block_limit,
+            MAX_VOTE_UNITS,
+        );
+
+        info!(
+            "Slot: {}. Cost Limits Reset. Bundle: {}, TX: {}",
+            working_bank.slot(),
+            self.current_bundle_block_limit,
+            self.current_tx_block_limit,
+        );
+    }
+
+    fn bundle_block_limit(&self) -> u64 {
+        self.current_bundle_block_limit
+    }
+
+    fn tx_block_limit(&self) -> u64 {
+        self.current_tx_block_limit
+    }
+
+    fn update_reserved_cost(&mut self, working_bank: &Arc<Bank>) {
+        if self.current_tx_block_limit != self.current_bundle_block_limit
+            && working_bank.tick_height() > working_bank.ticks_per_slot() * 4 / 5
```

Comment:
nit: saturating math

Reply:
also nit: ticks_per_slot shouldn't change so could set this on init

Reply:
so at worse, this will kick in 80% through slot + 10ms -> 400 * 0.8 =330ms through slot